Blog: Why use feature flags as a service? (#6650)
PostHog/posthog.com: contents/blog/feature-flags-as-a-service.md

- Speed and resiliency optimizations like local evaluation and caching.
- Logging and analytics.

## Why use a feature flags as a service provider?

The benefit of using a feature flag service is similar to the benefit of using other external services. You get the benefits of feature flags while (theoretically) minimizing their costs. This enables you to focus on building the core functionality that creates value in your business.

> **Aren’t feature flags as a service expensive?** Although you don't pay someone else to build your own feature flag service or use an open source version, you still pay to implement, maintain, and host it. This time, energy, and focus can cost much more than the dollar amount you pay a service provider.

With feature flags as a service, you gain more confidence in:

- **Usability.** The service provides a centralized location and UI to manage flags.

- **Reliability.** The service takes care of the infrastructure, redundancy, and speed.

- **Interoperability.** The service connects to any part of your application (SDKs) and other external services.

In fewer words, you pass your potential problems off to them. 

> ![Problems](../images/blog/feature-flags-as-a-service/problems.png)

Author: Meme template 🤔

> ![Problems](../images/blog/feature-flags-as-a-service/problems.png)

Author: https://github.com/PostHog/posthog.com/issues/6693

## Deciding whether to build or buy

> Like other services, the goal of a feature flag service is to help you build a successful product. Because of this, you should build instead of buy when this is true, such as when your use case is:

Reviewer: This paragraph is a bit verbose. Can we cut it down a bit?
```suggestion
Ultimately, the decision to build or buy is determined by which approach will enable you to build successful products. Thus, you should build instead of buy when your use is:
```

1. Unique enough to require a specific implementation tied to your core competency. For example, a crypto app requiring on-chain data for targeting.

> 2. Extremely simple. For example, you only need to toggle a feature on or off at build time.

Reviewer: ```suggestion
2. Extremely simple. For example, you only need to toggle a single feature on or off.
```

    if feature_flag_service.is_enabled("new_feature"):
        print("New feature enabled")
```

This enables you to simply toggle on or off features from a centralized location.

## The challenge of rolling your own feature flag service

The big problem with this simple service is that the values are hard coded and require a redeploy to change. It isn’t remotely configurable, meaning it can’t change at runtime and doesn’t provide many benefits over using `if` statements where needed.

The path to solving this is an increasingly complicated one, each step with its downsides:

1. Use a config to store the flag values. This still requires a redeploy to change. Read more in [Feature flags vs configuration: Which should you choose?](/blog/feature-flags-vs-configuration)
2. Store flag values in your database. This requires an admin panel or root access to modify the values and puts stress on your infrastructure. Database issues also impact the resiliency of flags.
3. Write an external service or use a library to manage and evaluate feature flags. New dependencies creates a bunch of maintenance and optimization work to ensure they remain fast and bug-free. 

> As is apparent, rolling your own in-house feature flag service becomes increasingly complicated. We haven’t even touched on adding logic for targeting, caching for speed, or resiliency for parts of the service going down. All of these make you more likely to introduce bugs and create tech debt. 

Reviewer: I think the first sentence is a bit repetitive of the opening paragraph of this section
```suggestion
We haven’t even touched on adding logic for targeting, caching for speed, or resiliency for parts of the service going down. All of these make you more likely to introduce bugs and create tech debt. 
```

> > **What does "roll your own" mean?** The phrase "rolling your own X" refers to developing, implementing, and using your own version of a common service such as authentication, bootloader, or in this case, feature flags.

Reviewer: I dont think this is needed. The phrase "roll your own" is well understood. Or alternatively, if we have to explain a phrase, we shouldnt use it

Author: Removed

A "feature flags as a service" provider is an external application that provides all the functionality for implementing and using feature flags. It is a centralized location to create, manage, evaluate, and monitor your feature flags. This enables remote configuration and the best practice of decoupling deployment from release.

Using feature flags as a service integrates with your app like other external services to provide a simple interface with. They create SDKs or an API to evaluate flags and get values. For example, in PostHog, a [feature flag call](/docs/feature-flags/adding-feature-flag-code) is as simple as this:

<MultiLanguage>

```js
import posthog from 'posthog-js'

posthog.init('<ph_project_api_key>', { api_host: '<ph_instance_address>' })

posthog.onFeatureFlags(function () {
    if (posthog.isFeatureEnabled('flag-key')) {
        // do something
    }
})
```

```python
from posthog import Posthog

posthog = Posthog('<ph_project_api_key>', host='<ph_instance_address>')

is_my_flag_enabled = posthog.feature_enabled('flag-key', 'distinct_id_of_your_user')
```

</MultiLanguage>

Feature flag services also contain features for complicated use cases including: 

> - Target feature flags to specific actors or situations to enable percentage rollouts, [betas](/tutorials/public-beta-program), [testing in production](/blog/testing-in-production), and more.
> - Multi-variant flags that enable A/B testing.
> - Integrations with other external services like product analytics, CDPs, and automations.
> - Speed and resiliency optimizations like local evaluation and caching.

Reviewer: ```suggestion
- Target feature flags to specific actors or situations to enable percentage rollouts, [betas](/tutorials/public-beta-program), [testing in production](/blog/testing-in-production), and more.
- Multi-variant flags that enable A/B testing.
- Integrations with other external services like product analytics, CDPs, and automations.
- Speed and resiliency optimizations like local evaluation and caching.
- Logging and analytics.
```

## Why use a feature flags as a service provider?

The benefit of using a feature flag service is similar to the benefit of using other external services. You get the benefits of feature flags while (theoretically) minimizing their costs. This enables you to focus on building the core functionality that creates value in your business.

> **Aren’t feature flags as a service expensive?** Although you aren’t paying anything to roll your own or use an open source version, you are still paying for the hosting, implementing, and maintaining of it. This time and energy can cost much more than the dollar amount you pay a service provider.

> With a feature flags as a service, you worry less about:
> 
> - **Usability.** The service provides a centralized location and UI to manage flags.
> - **Reliability.** The service takes care of the infrastructure, redundancy, speed, and reliability.
> - **Interoperability.** The service connects to any part of your application (SDKs) and other external services.

Reviewer: I think we should make this sentence positive instead of negative e.g.

```suggestion
With a feature flags as a service, you gain more confidence in:

- **Usability.** The service provides a centralized location and UI to manage flags.
- **Reliability.** The service takes care of the infrastructure, redundancy, speed, and reliability.
- **Interoperability.** The service connects to any part of your application (SDKs) and other external services.
```

> In fewer words, you pass your potential problems off to them. 

Reviewer: Dont think this is needed. We've made our point clearly enough above
```suggestion
```

Author: I want to keep this to set up my image below, it is kind of like a caption for it.

![Problems](../images/blog/feature-flags-as-a-service/problems.png)

> Using feature flags as a service doesn’t mean everything is perfect. There are often issues with the worry areas. Design can be poor, services can go down, and integrations can disappoint. 

Reviewer: ```suggestion
Using feature flags as a service doesn’t mean everything is perfect and there are other issues that may arise. Design can be poor, services can go down, and integrations can disappoint. 
```

> What using feature flags as a service means is that these issues are largely off your plate. You can focus on building a great product, and that is what a [feature flag as a service provider like PostHog](/feature-flags) is all about supporting.

Reviewer: I think this is repetitive of everything we've already covered
```suggestion
What using feature flags as a service means is that these issues are largely off your plate. You can focus on building a great product, and that is what a [feature flag as a service provider like PostHog](/feature-flags) is all about supporting.
```

Author: Remove/rewrote

> ## Further reading

Reviewer: Feels like the blog ended a bit abruptly. I think we could include a "Next steps" section or something similar. What do some of the other build vs buy blogs cover that we can include?

Author: Added a conclusion on making a decision.
